### PR TITLE
feat: implement the password change feature

### LINKED ISSUE #773: Implement the Change password feature
We started a change password feature, and that can be finished now.
Check ChangePasswordForm for the initial idea.
Update the component according to new Testnet standards.

On the Settings page add a Change password button.
Onclick the Form for change password appears with a new Save password button.
Make the necesarry validations for password checks (already present when registering for an account).
On Save password click, the new password is saved into the database.

### PR DESCRIPTION
**Pull request by trupakufi**
closes #773 

### Implement The Password Change Feature

The change password feature at the settings page, with the front and backend implementaions. 

### Changes

Changes we added listed below:

- add change password route on app.ts to handle the request.
- create the controller function changePassword on the userController to handle the request
- create changePassword on the user service to verify the password and set the new password
- create the user service api changePassword function to send the request from frontend to backend
- move and create more changePasswordSchema to validate the data both from frontend and backend
- organize the imports in some files.
- have some fun 😄 

@aosccode

### CODE REVIEW

**Comment by trupakufi** [comment]
@Tymmmy @rico191013 Can you check how is it ?

**Review by trupakufi — COMMENTED** [review]
It seems well.

File: packages/wallet/backend/src/user/service.ts [hunk 1]
```diff
@@ -1,19 +1,26 @@
...
+  email: string
+  newPassword: string
+  oldPassword: string
+}
+
 interface IUserService {
   create: (args: CreateUserArgs) => Promise<User>
   getByEmail(email: string): Promise<User | undefined>
   getById(id: string): Promise<User | undefined>
+  changePassword(ags: ChangePasswordArgs): Promise<void>
```
Comment (Tymmmy) [inline packages/wallet/backend/src/user/service.ts:23]:
```suggestion
  changePassword(args: ChangePasswordArgs): Promise<void>
```

File: packages/wallet/frontend/src/lib/api/user.ts [hunk 2]
```diff
@@ -255,6 +278,23 @@ const createUserService = (): UserService => ({
...
+        .post('change-password', {
+          json: args
+        })
+        .json<SuccessResponse>()
+
+      return response
+    } catch (error) {
+      return getError<ChangePasswordArgs>(
+        error,
+        'We could not update your passowrd. Please try again.'
```
Comment (Tymmmy) [inline packages/wallet/frontend/src/lib/api/user.ts:293]:
```suggestion
        'We could not update your password. Please try again.'
```

File: packages/wallet/frontend/src/components/settings/ChangePasswordForm.tsx [hunk 3]
```diff
@@ -34,8 +19,20 @@ export const ChangePasswordForm = () => {
...
-          console.log(data)
+        onSubmit={async (data) => {
+          const response = await userService.changePassword(data)
+
+          if (response.success) {
+            openDialog(
+              <SuccessDialog
+                onClose={closeDialog}
+                title="Password updated sucessfully."
+                content="Please check your inbox. Click on the provided link and reset your password."
```
Comment (Tymmmy) [inline packages/wallet/frontend/src/components/settings/ChangePasswordForm.tsx:30]: There will be no email sent to change password, the form will appear in the Settings page.
As I mentioned there is a button - Change password, and then the form appears in the page with the 3 inputs.
And a new button of Save.

**Comment by Tymmmy** [comment]
I need something like this to appear in the Settings page on a click of a button - Change password:
image

File: packages/wallet/frontend/src/components/settings/ChangePasswordForm.tsx (hunk 3 above)
Comment (trupakufi) [reply packages/wallet/frontend/src/components/settings/ChangePasswordForm.tsx]: Allright. I change it.

**Comment by trupakufi** [comment]
@Tymmmy I added it, I forgot to uncomment, now can you review it ?

**Comment by trupakufi** [comment]
@Tymmmy I added it, I forgot to uncomment, now can you review it ?

**Comment by trupakufi** [comment]
@Tymmmy

**Comment by trupakufi** [comment]
@Tymmmy I added it, I forgot to uncomment, now can you review it ?

**Comment by Tymmmy** [comment]
So...please read the requirements, and what I have been writing carefully :D. The screenshot I showed you, I mentioned that was from a previous start of the implementation, that was not finished. You can reuse that, no problem, BUT make sure it is up to date with newer forms, because it is old code, check for example the CreatePaymentPointerDialog.tsx form. 

AND as the story mentions:
"On the Settings page add a Change password button.
Onclick the Form for change password appears with a new Save password button."
And as I mentioned above in a few comments: "I need something like this to appear in the Settings page **on a click of a button** - Change password:"

This means:
- User goes to the settings page, sees the normal settings page, and a new button of Change Password
- User clicks on the button, the Change password form appears (as it appears right now without a click of a button)
- User has the possibility to change password by clicking on Save button
- User also has the possibility to hide the Change password form again by clicking on a Cancel button

So, again, please add a button to show hide Change password form in Settings. And right now even if the form appears and I add the input data, nothing happens on form submit. Nothing happens on Change Password click. So please check that also.

**Comment by devcer** [comment]
Thank you for Contributing to Interledger during Hacktoberfest. If you have any accepted PRs, You can claim your Swag [here](https://community.interledger.org/devcer/hacktoberfest-swag-redemption-claim-your-cool-gear-2152). Last date to fill up the form is December 10th, 2023. @trupakufi

**Comment by Tymmmy** [comment]
PR will close, we will revisit the story in another branch

**Pull request closed without merging**